web/src/features/map/Map.tsx

```diff
@@ -153,14 +156,54 @@ export default function MapPage({ onMapLoad }: MapPageProps): ReactElement {
 
   useEffect(() => {
     // Run on first load to center the map on the user's location
-    if (!map || isError || !isFirstLoad) {
+    if (!map || isError || !isFirstLoad || !isSourceLoaded || !data) {
       return;
     }
     if (data?.callerLocation && !selectedZoneId) {
       map.flyTo({ center: [data.callerLocation[0], data.callerLocation[1]] });
+
+      const handleIdle = () => {
+        if (
+          map.isSourceLoaded(ZONE_SOURCE) &&
+          map.areTilesLoaded() &&
+          data?.callerLocation
+        ) {
+          const source = map.getSource(ZONE_SOURCE);
+          const layer = map.getLayer('zones-clickable-layer');
+          if (!source) {
+            console.error(`Source "${ZONE_SOURCE}" not found`);
+            return;
+          }
+          if (!layer) {
+            console.error('Layer "zones-clickable-layer" not found or not rendered');
+            return;
+          }
+          const zoneFeature = getZoneIdFromLocation(
+            map,
+            data?.callerLocation,
+            ZONE_SOURCE
+          );
+          if (zoneFeature) {
+            const zoneName = zoneFeature.properties.zoneId;
+            setUserLocation(zoneName);
+            console.log('User is in zone:', zoneName);
+          }
+          map.off('idle', handleIdle);
+        }
+      };
       setIsFirstLoad(false);
+      map.on('idle', handleIdle);
     }
-  }, [map, isSuccess, isError, isFirstLoad, data?.callerLocation, selectedZoneId]);
+  }, [
+    map,
+    isSuccess,
+    isError,
+    isFirstLoad,
+    data,
```

Comment:
I think we need to be careful here and either deconstruct the data object first or explicitly list the callerLocation as the dependency like before. In theory the callerLocation can change without the data object changing.